### PR TITLE
Deployment of OCS with templates

### PR DESCRIPTION
**Pull request by clacroix12**

### CODE REVIEW

File: utility/aws.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                            'Value': name,
+                        },
+                    ]
+                },
+            ]
+        )
+        logger.debug("Response of volume creation: %s", volume_response)
+        volume = self.ec2_resource.Volume(volume_response['VolumeId'])
+        for x in range(timeout):
+            volume.reload()
```
Comment (vasukulkarni) [inline utility/aws.py]: Can we print the volume state once here?

File: utility/aws.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    @property
+    def ec2_client(self):
+        """ Property for ec2 client
+
+        Returns:
+            boto3.client: instance of ec2
+        """
+        if not self._ec2_client:
+            self._ec2_client = boto3.client('ec2')
```
Comment (vasukulkarni) [inline utility/aws.py]: When I was directly invoking the script via gitlab, I noticed that it asked me for region to be set either in .aws/env/or in this call. I think we can set it here with default region.

**Comment by vasukulkarni** [comment]
Overall looks good to me, didn't bother much with templates since this is a first cut. Thank you guys!

File: setup.py [hunk 3]
```diff
@@ -21,8 +21,9 @@
         'paramiko==2.4.2',
         'pyyaml>=4.2b1',
         'jinja2==2.10.1',
-        'kubernetes',
+        'kubernetes<9.0.0,>=8.0.0',
```
Comment (vasukulkarni) [inline setup.py]: Can we add comment for the reasoning behind <9 >8

File: tests/test_ocs_basic_install.py [hunk 4]
```diff
@@ -181,3 +275,84 @@ def load_config_data(data_path):
     """
     with open(data_path, "r") as data_descriptor:
         return yaml.load(data_descriptor, Loader=yaml.FullLoader)
+
+
+def create_rook_resource(template_name, rook_data, cluster_path):
```
Comment (petr-balogh) [inline tests/test_ocs_basic_install.py]: I would rename this function to something like create_oc_resource as it's about communicate with OCS, and would make the template path parametrized.

Move it to some place to be able to be easily used by others.

File: tests/test_ocs_basic_install.py [hunk 5]
```diff
@@ -181,3 +275,84 @@ def load_config_data(data_path):
...
+    """
+    Create a rook resource after rendering the specified template with
+    the rook data from cluster_conf.
+
+    Args:
+        template_name: name of the ocs-deployment config template.
```
Comment (petr-balogh) [inline tests/test_ocs_basic_install.py]: add types for params

File: utility/aws.py (hunk 1 above)
Comment (petr-balogh) [reply utility/aws.py]: we can print one more debug log here, I can add in next commit ;)

File: tests/test_ocs_basic_install.py [hunk 6]
```diff
@@ -181,3 +275,84 @@ def load_config_data(data_path):
...
+        rook_data.get(base_name, {})
+    )
+    cfg_file = os.path.join(cluster_path, template_name)
+    with open(cfg_file, "w") as f:
+        f.write(template)
+    log.info(f"Creating rook resource from {template_name}")
+    run_cmd(f"oc create -f {cfg_file}")
+
+
+def create_eb2_volumes(worker_pattern, size=100):
```
Comment (petr-balogh) [inline tests/test_ocs_basic_install.py]: rename to create_ebs_volume, I was probably confused with ec2 when I named it eb2. It's actually [EBS]( https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumes.html)

File: utility/aws.py (hunk 2 above)
Comment (petr-balogh) [reply utility/aws.py]: didn't hit such issue, probably some miss configured AWS in your home? We can add one in defaults and read it from there.

File: utility/aws.py (hunk 1 above)
Comment (petr-balogh) [reply utility/aws.py]: added in commit: https://github.com/red-hat-storage/ocs-ci/pull/23/commits/6659aa4f0c41c1972ffa9f7306ae8b7a8f4f52c1

File: utility/aws.py (hunk 2 above)
Comment (petr-balogh) [reply utility/aws.py]: Changed logic to have default region taken from default or config, check commit:  https://github.com/red-hat-storage/ocs-ci/pull/23/commits/6659aa4f0c41c1972ffa9f7306ae8b7a8f4f52c1

File: tests/test_ocs_basic_install.py [hunk 7]
```diff
@@ -150,6 +229,20 @@ def download_file(url, filename):
     assert r.ok
 
 
+def to_nice_yaml(a, indent=2, *args, **kw):
+    """Make verbose, human readable yaml"""
```
[the remark this answers is not in the document]
Comment (petr-balogh) [reply tests/test_ocs_basic_install.py]: Fixed and moved to different module in commit: https://github.com/red-hat-storage/ocs-ci/pull/23/commits/6659aa4f0c41c1972ffa9f7306ae8b7a8f4f52c1

File: tests/test_ocs_basic_install.py (hunk 5 above)
Comment (petr-balogh) [reply tests/test_ocs_basic_install.py]: done in https://github.com/red-hat-storage/ocs-ci/pull/23/commits/6659aa4f0c41c1972ffa9f7306ae8b7a8f4f52c1

File: tests/test_ocs_basic_install.py (hunk 6 above)
Comment (petr-balogh) [reply tests/test_ocs_basic_install.py]: fixed in https://github.com/red-hat-storage/ocs-ci/pull/23/commits/6659aa4f0c41c1972ffa9f7306ae8b7a8f4f52c1

File: setup.py (hunk 3 above)
Comment (clacroix12) [reply setup.py]: Yes, I'll add a comment in. This was to satisfy the version needed for the `openshift` dependency.

**Comment by vasukulkarni** [comment]
How about we redo the commit and keep commit history clean after all review comments are addressed? I dont think it qualifies for squash and merge

Ex: commit1: template-file1 - description about its role and how its used
      commit2: template-file2 - description about its role and how its used
      commit3: aws changes - 
and so on..

File: tests/test_ocs_basic_install.py (hunk 4 above)
Comment (clacroix12) [reply tests/test_ocs_basic_install.py]: Fixed

File: setup.py (hunk 3 above)
Comment (clacroix12) [reply setup.py]: Fixed

**Comment by clacroix12** [comment]
Cleaned up the duplicate commits and squashed several others. Please take another look and let me know if anything looks off.

File: tests/test_ocs_basic_install.py [hunk 8]
```diff
@@ -181,3 +254,28 @@ def load_config_data(data_path):
...
+        template_path,
+        rook_data.get(base_name, {})
+    )
+    cfg_file = os.path.join(cluster_path, template_name)
+    with open(cfg_file, "w") as f:
+        f.write(template)
+    log.info(f"Creating rook resource from {template_name}")
+    # TODO: logging this just for testing purposes, change to run_cmd
+    log.info(f"oc create -f {cfg_file}")
+>>>>>>> 2993ae6... Coady's changes for ocs deployment
```
Comment (vasukulkarni) [inline tests/test_ocs_basic_install.py]: ha this one, please :-)

Comment (clacroix12) [reply tests/test_ocs_basic_install.py]: i think this is an outdated version of the file

**Comment by vasukulkarni** [comment]
I would like to merge this latest by Monday, if there is anything that is blocking please add your comment or else It will be nice if we can take that up as an issue in next PR

File: tests/test_ocs_basic_install.py (hunk 8 above)
Comment (vasukulkarni) [reply tests/test_ocs_basic_install.py]: That is correct, Thanks

**Pull request merged**